amethyst_ui/src/action_components.rs

```diff
@@ -0,0 +1,58 @@
+use amethyst_audio::SourceHandle;
+use amethyst_core::specs::prelude::{Component, DenseVecStorage};
+use amethyst_renderer::TextureHandle;
+
+/// When this component is added to a UI element with a `UiImage`
+/// it will change that image based on mouse interaction.
+/// Requires `MouseReactive`.
+pub struct OnUiActionImage {
+    /// Default image
+    pub(crate) normal_image: TextureHandle,
+    /// Image used when the mouse hovers over this element
+    pub(crate) hover_image: Option<TextureHandle>,
+    /// Image used when element is pressed
+    pub(crate) press_image: Option<TextureHandle>,
+}
+
+impl OnUiActionImage {
+    /// A constructor for this component
+    pub fn new(
+        normal_image: TextureHandle,
+        hover_image: Option<TextureHandle>,
+        press_image: Option<TextureHandle>,
+    ) -> Self {
+        Self {
+            normal_image,
+            hover_image,
+            press_image,
+        }
+    }
+}
+
+impl Component for OnUiActionImage {
+    type Storage = DenseVecStorage<Self>;
+}
+
+/// When this component is added to a UI element
+/// it will play sounds based on mouse interaction.
+/// Requires `MouseReactive`.
+pub struct OnUiActionSound {
+    /// Sound made when this button is hovered over
+    pub(crate) hover_sound: Option<SourceHandle>,
+    /// Sound made when this button is pressed.
+    pub(crate) press_sound: Option<SourceHandle>,
```

Comment:
release sound too? :)

Reply:
Sure, sounds good!

Reply:
Ah yeah about that, when I come back to this I find that this one is a bit of a misnomer.  We're already using this sound as the release sound.  I'll fix the name.

Reply:
Ya, but can't you have one for click start and one for click end (and one for click hold? :D )

Reply:
I haven't seen any game UI that does those things, I'm not sure there's a strong use case for it.

Reply:
age of empires 3 does this. It does add a lot of feel to the buttons when there's a release sound that is different from the press sound.